Give LineEdit/TextEdit a custom color for font while uneditable

GlaceGwyneth opened the pull request:
By default an uneditable LineEdit's text has some transparency hardcoded.
This change adds a custom color to LineEdit for adjusting the font when editable is unchecked.
It also leaves the default behavior (0.5 alpha) unchanged.

Fixes #29814

GlaceGwyneth committed:
Give LineEdit a custom font color when un-editable

By default the LineEdit's text when editable is unchecked had some transparency hardcoded.
This change adds a custom color to LineEdit for adjusting the font when editable is off.
Addresses issue 29814

YeldhamDev commented:
I believe `TextEdit` also does that when in read-only mode. This change should be made there as well then.

GlaceGwyneth committed:
Give TextEdit a custom color for font when read only is set

Previously there was some transparency hard coded into TextEdit when in read only mode.
This change adds a custom color for adjusting the font in read only mode.
It also applies when syntax highlighting is on.

GlaceGwyneth commented:
Hopefully I didn't miss anything while testing...
One thing I noticed:
The order of the custom colors in editor isn't the same as the order in the code, so the readonly color is way off at the bottom by itself, but I'm not sure how to control this?

YeldhamDev commented:
Thank you for the changes. And happy first contribution!

YeldhamDev commented:
From my quick look in the editor, it seems that code order doesn't affect constants, so you're in the green on that.

However, Travis detect some errors. I will comment on the lines that they were found shortly.

YeldhamDev commented on scene/gui/text_edit.cpp:
```diff
@@ -675,8 +675,8 @@ void TextEdit::_notification(int p_what) {
-			Color color = cache.font_color;
-			color.a *= readonly_alpha;
+			Color color = readonly ? cache.font_color_readonly : cache.font_color;
+			
```
Unneeded extra line.

YeldhamDev commented:
The `readonly_alpha` variable found in line 665 is now unused and should be removed.
(which I would have loved to have commented directly on it, but GitHub doesn't allow commenting on lines too far away from modifications because "reasons" and- this is not the proper place to rant about that isn't it?)

YeldhamDev commented:
That should be it. I recommend getting Godot's [git hooks](https://github.com/godotengine/godot/tree/master/misc/hooks), so clang will alert you when breaking formatting.

akien-mga commented on scene/resources/default_theme/default_theme.cpp:
```diff
@@ -440,6 +440,7 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
 	theme->set_color("completion_font_color", "TextEdit", Color::html("aaaaaa"));
 	theme->set_color("font_color", "TextEdit", control_font_color);
 	theme->set_color("font_color_selected", "TextEdit", Color(0, 0, 0));
+	theme->set_color("font_color_read_only", "TextEdit", Color(control_font_color.r, control_font_color.g, control_font_color.b, 0.5f));
```
The property is named `readonly` in one word, so it's probably best if the color follows the same spelling.

akien-mga commented on scene/gui/line_edit.cpp:
```diff
@@ -724,15 +724,14 @@ void LineEdit::_notification(int p_what) {
 			// draw placeholder color
 			if (using_placeholder)
 				font_color.a *= placeholder_alpha;
-			font_color.a *= disabled_alpha;
```
I think you can remove the `disabled_alpha` variable which is now only used once at line 740 (you can use a ternary operator there instead to set the alpha).

GlaceGwyneth replied:
I felt a bit weird about the ternary bit, does it look okay?

akien-mga replied:
I would have simplified to `is_editable() ? 0.9 : 0.45`, but it's also ok as is.

akien-mga reviewed (approved):
Looks good, and it seems like you made sure the new properties will produce the same visual effect by default, so it should not break compatibility.

I left a couple comments, otherwise it should be good to merge.

akien-mga commented:
Thanks!

Shadowblitz16 commented:
@akien-mga was this reverted?
there is no option for disabled text color for line edit
![image](https://user-images.githubusercontent.com/2453707/98450946-1d446f80-20f6-11eb-9fa1-e9d753bb5e96.png)

akien-mga commented:
I see "Read Only" among colors in your screenshot...

Shadowblitz16 commented:
@akien-mga  the thing is its not changing anything

bruvzg commented:
Color name is `font_color_uneditable`, and it's available if you add the item to the theme `Edit Theme->Add Item`, select `Type: LineEdit`, `Data Type: Color`, `Name: font_color_uneditable`. There's no color with name `read_only`, you are probably using a theme created in the old engine version.
